bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/discovery/GreeDeviceFinder.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.gree.internal.discovery;
+
+import static org.openhab.binding.gree.internal.GreeBindingConstants.*;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.gree.internal.GreeCryptoUtil;
+import org.openhab.binding.gree.internal.GreeException;
+import org.openhab.binding.gree.internal.gson.GreeScanReponsePack4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeScanRequest4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeScanResponse4GsonDTO;
+import org.openhab.binding.gree.internal.handler.GreeAirDevice;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * The GreeDeviceFinder provides functionality for searching for GREE Airconditioners on the network and keeping a list
+ * of found devices.
+ *
+ * @author John Cunha - Initial contribution
+ * @author Markus Michels - Refactoring, adapted to OH 2.5x
+ */
+@NonNullByDefault
+public class GreeDeviceFinder {
+    private final Logger logger = LoggerFactory.getLogger(GreeDeviceFinder.class);
+
+    protected final InetAddress mIPAddress;
+    protected HashMap<String, GreeAirDevice> mDevicesHashMap = new HashMap<>();
+
+    public GreeDeviceFinder() {
+        mIPAddress = InetAddress.getLoopbackAddress(); // dummy
+    }
+
+    public GreeDeviceFinder(String broadcastAddress) throws UnknownHostException {
+        mIPAddress = InetAddress.getByName(broadcastAddress);
+    }
+
+    public void scan(Optional<DatagramSocket> socket, boolean scanNetwork) throws GreeException {
```

Comment:
I want to get rid of that @Nullable crap as much as possible because it always leads into "potential null access" etc. I will keep it

Reply:
The nullable crap only becomes an annoyance when it comes to fields, which is due to it trying to be thread-safe. The null checker works perfectly fine on local variable / parameters where only the running thread has access to it.

Reply:
I tried it once more and I like the Optional version way more, otherwise
- instance variable clientSocket must be Nullable
- all parameters in the GreeHandler functions need to be changed to nullable
- all parameters in GreeAirDrvice too
- then we run in "potential null pointer access", which adds a bunch of "if (clientSocket == null")...
all that, because in theory the variable can be null, which is never the case when initialization was successful

For this I want to stay with the Optional approach.

Reply:
@cpmeister Do you agree after looking to the latest code?